test/services/database/post_upvote_repository_test.dart

```diff
@@ -0,0 +1,302 @@
+import "dart:math";
+
+import "package:cloud_firestore/cloud_firestore.dart";
+import "package:collection/collection.dart";
+import "package:fake_cloud_firestore/fake_cloud_firestore.dart";
+import "package:flutter_test/flutter_test.dart";
+import "package:proxima/models/database/post/post_data.dart";
+import "package:proxima/models/database/post/post_firestore.dart";
+import "package:proxima/models/database/post/post_id_firestore.dart";
+import "package:proxima/models/database/upvote_state.dart";
+import "package:proxima/models/database/user/user_id_firestore.dart";
+import "package:proxima/services/database/post_repository_service.dart";
+import "package:proxima/services/database/post_upvote_service.dart";
+
+void main() {
+  late FakeFirebaseFirestore firestore;
+  late PostRepositoryService postRepository;
+  late PostUpvoteRepositoryService postUpvoteRepository;
+
+  Future<int> getUpvoteCount(PostIdFirestore postId) async {
+    final post = await postRepository.getPost(postId);
+    return post.data.voteScore;
+  }
+
+  const ownerId = UserIdFirestore(value: "owner_id");
+  final postData = PostData(
+    ownerId: ownerId,
+    title: "",
+    description: "",
+    publicationTime: Timestamp.fromMillisecondsSinceEpoch(0),
+    voteScore: 0,
+  );
+  const postLocation = GeoPoint(40, 20);
+
+  setUp(() async {
+    firestore = FakeFirebaseFirestore();
+    postRepository = PostRepositoryService(firestore: firestore);
+    postUpvoteRepository = PostUpvoteRepositoryService(firestore: firestore);
+  });
+
+  int stateToValue(UpvoteState state) {
+    switch (state) {
+      case UpvoteState.none:
+        return 0;
+      case UpvoteState.upvoted:
+        return 1;
+      case UpvoteState.downvoted:
+        return -1;
+    }
+  }
+
+  Future<void> assertPostUpvoteState(
+    UserIdFirestore userId,
+    PostIdFirestore postId,
+    UpvoteState expectedState, {
+    testScore = true,
+  }) async {
+    final upvoteState = await postUpvoteRepository.getUpvoteState(
+      userId,
+      postId,
+    );
+    expect(upvoteState, expectedState);
+
+    if (testScore) {
+      final expectedScore = stateToValue(expectedState);
+      expect(await getUpvoteCount(postId), expectedScore);
+    }
+  }
+
+  group(
+    "Operation combination work for a single user with a single post",
+    () {
+      late PostFirestore post;
+      const userId = UserIdFirestore(value: "user_id");
+
+      setUp(() async {
+        final id = await postRepository.addPost(postData, postLocation);
+        post = await postRepository.getPost(id);
+      });
+
+      void testActions(List<UpvoteState> states) async {
```

Comment:
Is there any reason why this function is marked as async ?

Reply:
No indeed, fixed in eadff8fc2ad6c2e54b8b9c12de364be006f46d9c.